YARCH/Sources/Modules/CatalogDetails/CatalogDetailsInteractor.swift

```diff
@@ -0,0 +1,91 @@
+                    let catalogResult: CatalogDetails.FetchDetails.Response = .init(result: .success(coinModel, data))
+                    DispatchQueue.main.async {
+                        self?.presenter.presentFetchedDetails(response: catalogResult)
+                    }
+                })
+                let catalogResult: CatalogDetails.FetchDetails.Response = .init(result: .success(coinModel, nil))
+                self?.presenter.presentFetchedDetails(response: catalogResult)
+            case let .failure(error as CatalogDetailsError):
+                let response: CatalogDetails.FetchDetails.Response = .init(result: .failure(id: request.coinId, error))
+                self?.presenter.presentFetchedDetails(response: response)
+            case let .failure(error):
+                let error = CatalogDetailsError.networkError(error.localizedDescription)
+                self?.presenter.presentFetchedDetails(response: CatalogDetails.FetchDetails.Response(result: .failure(id: request.coinId, error)))
+            }
+        }
+    }
+
+    func downloadImage(_ urlString: String, completion: @escaping ((Data) -> Void)) {
+        imageProvider.loadImageData(urlString, completion: completion)
+    }
+
+    // MARK: Open External Link
+
+    func openExternalLink(request: CatalogDetails.OpenExternalLink.Request) {
+        provider.fetchDetails(coinId: request.coinId) { [weak self] result in
+            switch result {
+            case let .success(coinModel):
+                self?.performOpenExternalLink(coinModel, externalLinkType: request.type)
+            case let .failure(error as CatalogDetailsError):
+                let viewModel: CatalogDetails.FetchDetails.Response = .init(result: .failure(id: request.coinId, error))
+                self?.presenter.presentFetchedDetails(response: viewModel)
+            case let .failure(error):
+                let error = CatalogDetailsError.networkError(error.localizedDescription)
+                self?.presenter.presentFetchedDetails(response: CatalogDetails.FetchDetails.Response(result: .failure(id: request.coinId, error)))
+            }
+        }
+    }
+
+    func performOpenExternalLink(_ coinModel: CoinSnapshotFullModel, externalLinkType: CatalogDetails.OpenExternalLink.ExternalLinkType) {
+        var response: CatalogDetails.OpenExternalLink.Response = .init(result: .failure(CatalogDetailsError.otherLogicError))
+        var url: URL? = nil
+        switch externalLinkType {
+        case .website:
+            url = coinModel.website.extractURLs().first
+        case .twitter:
+            url = URL(string: "\(Configuration.twitterBaseUrlString)\(coinModel.twitter)")
+        }
+        if let url = url {
+            response = .init(result: .success(url))
+        } else {
+            response = .init(result: .failure(CatalogDetailsError.externalURLError))
+        }
+        presenter.presentOpenExternalLink(response: response)
```

Comment:
May be refactored to: 
```
let response: ...
defer { presenter.present(response: response) }
...
guard let url = url {
  response = .init(result: .failure(CatalogDetailsError.externalURLError))
}
response = .init(result: .success(url))
```

Reply:
Thx. Unfortunately, we should use ugly `unwrappedVar` in `guard let ...` statement  in order to make compiler happy...  
<img width="795" alt="2018-03-04 19 05 09" src="https://user-images.githubusercontent.com/9966597/36947583-20da8a28-1fdf-11e8-9fb5-7c758f6938d4.png">